src/storage/bg_task/compact_segments_task.cpp

```diff
@@ -0,0 +1,246 @@
+                    remapper.AddMap(segment_id, block_entry->block_id(), row_begin, new_row_id);
+                    read_offset = row_begin + read_size;
+                };
+
+                if (read_size + current_block_ptr->row_count() > current_block_ptr->row_capacity()) {
+                    SizeT read_size1 = current_block_ptr->row_capacity() - current_block_ptr->row_count();
+                    block_entry_append(row_begin, read_size1);
+                    row_begin += read_size1;
+                    read_size -= read_size1;
+                    if (current_block_entry.get() != nullptr) {
+                        new_segment->AppendBlockEntry(std::move(current_block_entry));
+                    }
+                    BlockID next_block_id = new_segment->block_entries().size();
+                    current_block_entry = BlockEntry::NewBlockEntry(new_segment.get(), next_block_id, 0, column_count, txn_);
+                    current_block_ptr = current_block_entry.get();
+                }
+                block_entry_append(row_begin, read_size);
+            }
+        }
+    }
+    if (current_block_entry.get() != nullptr && current_block_entry->row_count() > 0) {
+        new_segment->AppendBlockEntry(std::move(current_block_entry));
+    }
+
+    // TODO: Increase row count here is not good
+    for (auto &block_entry : new_segment->block_entries()) {
+        new_segment->IncreaseRowCount(block_entry->row_count());
+    }
+
+    return new_segment;
+}
+
+// Set old segment deprecated and add compact_wal_cmd, import_wal_cmd
+void CompactSegmentsTask::SaveSegmentsData(Vector<Pair<SharedPtr<SegmentEntry>, Vector<SegmentID>>> &&segment_data) {
+    TxnTimeStamp begin_ts = txn_->BeginTS();
+    for (auto &[new_segment, old_segment_ids] : segment_data) {
+        new_segment->FlushData();
+
+        auto *txn_store = txn_->GetTxnTableStore(table_entry_);
+        txn_store->Import(new_segment);
+
+        const auto &block_entries = new_segment->block_entries();
+        SizeT block_entries_size = block_entries.size();
+        u16 last_block_row_count = block_entries.back()->row_count();
+        String db_name = *table_entry_->GetDBName();
+        String table_name = *table_entry_->GetTableName();
+        txn_->AddWalCmd(MakeShared<WalCmdImport>(db_name,
+                                                 table_name,
+                                                 *new_segment->segment_dir(),
+                                                 new_segment->segment_id(),
+                                                 block_entries_size,
+                                                 DEFAULT_BLOCK_CAPACITY, // TODO: store block capacity in segment entry
+                                                 last_block_row_count));
+
+        for (SegmentID segment_id : old_segment_ids) {
+            auto *segment_entry = table_entry_->segment_map().at(segment_id).get();
+            segment_entry->SetDeprecated(begin_ts);
```

Comment:
These deprecated segment_entry need to be recorded in txn store and WAL. Used in following two cases:
1. Rollbacking the txn will remove the deprecated TS.
2. Replay WAL will deprecated the old segment_entry.